crates/nu-cli/src/completions/arg_value_dynamic_completion.rs

```diff
@@ -0,0 +1,57 @@
+use crate::completions::{
+    Completer, CompletionOptions, SemanticSuggestion, completion_options::NuMatcher,
+};
+use nu_protocol::{
+    DeclId, Span,
+    engine::{ArgType, Stack, StateWorkingSet},
+};
+use reedline::Suggestion;
+
+pub struct ArgValueDynamicCompletion<'a> {
+    pub decl_id: DeclId,
+    pub arg_type: ArgType<'a>,
+}
+
+impl<'a> Completer for ArgValueDynamicCompletion<'a> {
+    fn fetch(
+        &mut self,
+        working_set: &StateWorkingSet,
+        stack: &Stack,
+        prefix: impl AsRef<str>,
+        span: Span,
+        offset: usize,
+        options: &CompletionOptions,
+    ) -> Vec<SemanticSuggestion> {
+        // the `prefix` is the value of a flag
+        // if user input `--foo abc`, then the `prefix` here is abc.
+        // the name of flag is saved in `self.flag_name`.
+        let mut matcher = NuMatcher::new(prefix, options);
+        let mut add_suggestion = |value: String| {
+            matcher.add_semantic_suggestion(SemanticSuggestion {
+                suggestion: Suggestion {
+                    value,
+                    description: None,
+                    span: reedline::Span {
+                        start: span.start - offset,
+                        end: span.end - offset,
+                    },
+                    append_whitespace: true,
+                    ..Suggestion::default()
+                },
+                kind: None,
+            });
+        };
+
+        let decl = working_set.get_decl(self.decl_id);
+        let mut stack = stack.to_owned();
+        if let Some(items) = decl
+            .get_dynamic_completion(working_set.permanent_state, &mut stack, &self.arg_type)
+            .unwrap_or_default()
+        {
+            for i in items {
+                add_suggestion(i);
+            }
+        }
+        matcher.results()
```

Comment:
I think it's a good idea to move the logic of

https://github.com/nushell/nushell/blob/18953d517eefdc6d5c1f85c4609fa75755b9e870/crates/nu-cli/src/completions/completer.rs#L715

to this place, helps with some common features, such as type-based completion, to be shared between positional and flag values.

We shouldn't do it within this PR, but would you please add a TODO note to `argument_completion_helper` for me?